web/src/filter.js

```diff
@@ -550,29 +550,57 @@ export class Filter {
         return this._can_mark_messages_read;
     }
 
+    // This function is used to check whether all the messages of the conversation
+    // are present or not of which a message is present in the view.
+    are_all_conversation_messages_of_a_message_in_view() {
+        const term_types = this.sorted_term_types();
+
+        if (_.isEqual(term_types, ["stream", "topic"])) {
+            return true;
+        }
+
+        if (_.isEqual(term_types, ["dm"])) {
+            return true;
+        }
+
+        if (_.isEqual(term_types, ["stream"])) {
+            return true;
+        }
+
+        if (_.isEqual(term_types, ["is-dm"])) {
+            return true;
+        }
+
+        if (_.isEqual(term_types, ["is-resolved"])) {
+            return true;
+        }
+
+        if (term_types.length === 1 && ["in-home", "in-all"].includes(term_types[0])) {
+            return true;
+        }
+
+        return false;
+    }
+
     // This is used to control the behaviour for "exiting search",
     // given the ability to flip between displaying the search bar and the narrow description in UI
     // here we define a narrow as a "common narrow" on the basis of
     // https://paper.dropbox.com/doc/Navbar-behavior-table--AvnMKN4ogj3k2YF5jTbOiVv_AQ-cNOGtu7kSdtnKBizKXJge
     // common narrows show a narrow description and allow the user to
     // close search bar UI and show the narrow description UI.
-    //
-    // TODO: We likely will want to rewrite this to not piggy-back on
-    // can_mark_messages_read, since that might gain some more complex behavior
-    // with near: narrows.
     is_common_narrow() {
-        // can_mark_messages_read tests the following filters:
+        // are_all_conversation_messages_of_a_message_in_view tests the following filters:
         // stream, stream + topic,
         // is:dm, dm,
         // is:resolved
-        if (this.can_mark_messages_read()) {
+        if (this.are_all_conversation_messages_of_a_message_in_view()) {
             return true;
         }
         // that leaves us with checking:
         // is: starred
-        // (which can_mark_messages_read does not check as starred messages are always read)
+        // (which are_all_conversation_messages_of_a_message_in_view does not check as starred messages are always read)
         // is: mentioned
-        // (for which can_mark_messages_read returns false, since they don't have conversation context)
+        // (for which are_all_conversation_messages_of_a_message_in_view returns false, since they don't have conversation context)
```

Comment:
I feel like writing a better comment here would be helpful with this refactor too.

Reply:
Updated the comment, do let me know if you want me to make anymore changes,